subsys/net/lib/mqtt/mqtt_transport_websocket.c

```diff
@@ -141,13 +141,22 @@ int mqtt_client_websocket_read(struct mqtt_client *client, u8_t *data,
 			       u32_t buflen, bool shall_block)
 {
 	s32_t timeout = K_FOREVER;
+	u32_t message_type;
+	int ret;
 
 	if (!shall_block) {
 		timeout = K_NO_WAIT;
 	}
 
-	return websocket_recv_msg(client->transport.websocket.sock,
-				  data, buflen, NULL, NULL, timeout);
+	ret = websocket_recv_msg(client->transport.websocket.sock,
+				 data, buflen, &message_type, NULL, timeout);
+	if (ret > 0) {
+		if (!(message_type & WEBSOCKET_FLAG_BINARY)) {
```

Comment:
Shouldn't `WEBSOCKET_FLAG_CLOSE` be handled separately? Otherwise, MQTT engine will not be aware that peer closed the connection. Socket based transport implementations simply return `0` in that case.

Reply:
Sure, that can be added here. I just wonder that if we receive CLOSE, then the underlying socket is probably closed anyway by the peer so the end result is probably the same. Anyway, I will send new version.

Reply:
[RFC 6455, chapter 7.1.2](https://tools.ietf.org/html/rfc6455#section-7.1.2):

> Once an endpoint has both sent and received a Close control frame, that endpoint SHOULD _Close the WebSocket Connection_ as defined in Section 7.1.1.